subprojects/build-plugins/src/main/kotlin/org/gradle/kotlin/dsl/build/plugins/KotlinDslJavaApiExtensionsPlugin.kt

```diff
@@ -0,0 +1,215 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.kotlin.dsl.build.plugins
+
+import org.gradle.api.Named
+import org.gradle.api.NamedDomainObjectContainer
+import org.gradle.api.Plugin
+import org.gradle.api.Project
+import org.gradle.api.file.FileTree
+import org.gradle.api.plugins.JavaPluginConvention
+import org.gradle.api.tasks.SourceSet
+import org.gradle.api.tasks.TaskProvider
+import org.gradle.jvm.tasks.Jar
+
+import org.gradle.kotlin.dsl.*
+
+import org.gradle.kotlin.dsl.build.tasks.GenerateKotlinDslApiExtensions
+import org.gradle.kotlin.dsl.build.tasks.GenerateParameterNamesIndexProperties
+
+import org.jetbrains.kotlin.gradle.plugin.KotlinSourceSet
+import org.jetbrains.kotlin.gradle.tasks.KotlinCompile
+
+
+class KotlinDslJavaApiExtensionsPlugin : Plugin<Project> {
+
+    override fun apply(project: Project): Unit = project.run {
+
+        apply<GeneratedSourcesLayoutPlugin>()
+        apply(plugin = "kotlin")
+
+        val kotlinDslApiExtensions = project.container(KotlinDslApiExtensionsSet::class.java) { name ->
+            KotlinDslApiExtensionsSet(project, name)
+        }
+
+        extensions.add(
+            typeOf<NamedDomainObjectContainer<KotlinDslApiExtensionsSet>>(),
+            "kotlinDslApiExtensions",
+            kotlinDslApiExtensions)
+
+        afterEvaluate {
+            kotlinDslApiExtensions.all { extensionsSet ->
+                configureProjectFor(extensionsSet)
+            }
+        }
+    }
+
+    private
+    fun Project.configureProjectFor(extensionsSet: KotlinDslApiExtensionsSet) =
+        createKotlinDslExtensionsSourceSetFor(extensionsSet.sourceSet.get()).let { extensionsSourceSet ->
+            registerParameterNamesIndexTaskFor(extensionsSet).let { parameterNamesTask ->
+                registerKotlinDslApiExtensionsTaskFor(extensionsSet, extensionsSourceSet, parameterNamesTask)
+            }
+        }
+
+    private
+    fun Project.createKotlinDslExtensionsSourceSetFor(sourceSet: SourceSet) =
+        java.sourceSets.create("${sourceSet.name}KotlinDslExtensions") {
+            it.compileClasspath += sourceSet.compileClasspath + sourceSet.output.classesDirs
+        }.also { extensionsSourceSet ->
+            dependencies {
+                (extensionsSourceSet.compileClasspathConfigurationName)(kotlin("stdlib-jdk8"))
+            }
+            tasks {
+                getByName<KotlinCompile>(extensionsSourceSet.getCompileTaskName("kotlin")) {
+                    kotlinOptions {
+                        jvmTarget = "1.8"
+                        freeCompilerArgs += "-Xjsr305=strict"
+                    }
+                }
+                getByName<Jar>("jar") {
+                    from(extensionsSourceSet.kotlin.sourceDirectories)
+                    from(extensionsSourceSet.output)
+                    dependsOn(extensionsSourceSet.classesTaskName)
+                }
+            }
+        }
```

Comment:
Do you also need to configure the "apielements" outgoing variants?
https://docs.gradle.org/current/userguide/java_library_plugin.html#sec:java_library_known_issues
```
configurations {
    "apiElements" {
        outgoing
            .variants
            .getByName("classes")
            .artifact(mapOf(
                "file" to compileKotlin.destinationDir,
                "type" to "java-classes-directory",
                "builtBy" to compileKotlin
            ))
    }
}
```